### PR TITLE
[ENH] Add carpet plot to outputs

### LINKED ISSUE #688: Add carpet plots to HTML reports
### Summary
It would be great to include carpet plots in the HTML reports. In a perfect world, those plots would be broken down by tissue type and would also have line plots attached with regressors of no interest, but those elements would require additional inputs, which we may not want to support.

### Additional Detail
I have been working on carpet plots in `nilearn`, so we could use their functions.

This stems from #450.

### PR DESCRIPTION
**Pull request by tsalo**
Closes #688.

Changes proposed in this pull request:

- Generate a set of carpet plots in a single figure and output to the figures directory.
- Add the carpet plots to the HTML report.
- Add new function (`tedana.io.denoise_ts`) to handle the denoising part.
- Drop unused output (`varexpl`) from `tedana.io.write_split_ts`.

To do:

- [x] Add info and image of carpet plots to our RTD documentation.
- [ ] Regenerate example report in [our demo](https://me-ica.github.io/tedana-ohbm-2020/). We actually might want to do this _after_ we've handled all of the current report-related PRs.

### CODE REVIEW

**Comment by tsalo** [comment]
@eurunuela would you be willing to help with the report integration? I'm not sure who else is familiar with the current report code (mostly because I didn't help with its development).

**Comment by eurunuela** [comment]
@tsalo I'll be happy to help. We're having our eastern holidays starting next week, I will try to catch up on my (many) side projects, this included. I'm adding it to my todo list.

**Comment by eurunuela** [comment]
I'm sorry it took me 2 months to get to this. I've just pushed some code that adds the carpet plots to the report. 

Let me know what you think @tsalo @emdupre

**Comment by emdupre** [comment]
Coming in very late (but before the artifacts expire !), here's what it looks like for me :

Screenshot 2021-06-11 at 14-22-30 tedana report

The only two suggestions I could make (not sure if we want to handle them here or in subsequent PRs):

- I know nilearn is working on grey-white mask generation. Could we include something like that to separate out tissue types ?
- Could we make the reports in a dropdown so you can choose which plot to view instead of viewing them stacked together ?

**Comment by tsalo** [comment]
> I know nilearn is working on grey-white mask generation. Could we include something like that to separate out tissue types?

I assumed that mask generation was from T1s. Will it work on the OC data?

> Could we make the reports in a dropdown so you can choose which plot to view instead of viewing them stacked together?

That sounds _awesome_. I have no clue how to do it though. Maybe @eurunuela could help with that?

**Comment by notZaki** [comment]
To show individual plots, it might be convenient to have the carpet plots split into individual files, with one carpet plot per file. Then the displayed image can be controlled with a radio button.

**Comment by tsalo** [comment]
That's a great point. I think that's the way to go. I can handle splitting the figures and adding them to the report, but I'll need someone else to figure out the radio button.

**Comment by tsalo** [comment]
Well, I've got the plots split up and I _think_ I may have messed up the layout in the report:

image

**Comment by jbteves** [comment]
@tsalo how do you feel about making that a new row instead of a new column? I feel like it's a little more organized to put it as a new row, and that might solve the weird whitespacing you're seeing.

**Comment by eurunuela** [comment]
I'll have a look at it next week @tsalo . If someone else wants to help with that, please go ahead.

**Comment by dowdlelt** [comment]
A small counter to having not stacked plots - I think it is interesting to scan down through a stack and see what artifacts go away, and those that remain. That is a little more difficult when one is switching between each. Though - I do agree with (what I think is...) the general sentiment - that a long stack is a bit much/too long/etc. If there are buttons that show specific images,  what about a "overwhelm me with data" button that shows all?

**Comment by eurunuela** [comment]
I have no idea why tests don't even run.

I think you'll like the new design for the carpet plots btw.

**Comment by tsalo** [comment]
@eurunuela I just ran it locally and it is _awesome_! For posterity, here's what it looks like:

image

(On an unrelated note, the vertical banding in the high-kappa data there is _crazy_ ☹️)

**Comment by tsalo** [comment]
I hoped that updating from `main` would fix the problem, but it doesn't seem to have done anything. The issue appears to be with `make`, even though this PR didn't touch the Makefile. @jbteves I think you know more about `make` than I do. Would you mind taking a look at the failing tests?

**Comment by notZaki** [comment]
It could be an `apt` issue. Refreshing the source list with `apt update` or similar might work.

I thought the banding was expected, as long as it is a global trend.

**Comment by tsalo** [comment]
I'm getting the same error in other tedana PRs and even NiMARE ones. Here's the error message for posterity:

> E: Package 'make' has no installation candidate

@notZaki do you know how to do that here?

**Comment by notZaki** [comment]
@tsalo Maybe try sprinkling some `apt-get update`s before all the `apt-get install`s in the circle-ci config and see if that solves it.[[Example]](https://github.com/notZaki/tedana/commit/4360af44570e42f173353a9a980cf1a018f89a28)

**Comment by eurunuela** [comment]
> I'm getting the same error in other tedana PRs and even NiMARE ones. Here's the error message for posterity:

If you're getting the same error on other repos too, could the issue be on CircleCI's end?

**Comment by tsalo** [comment]
What do you think about the GSR/MIR-denoised data? Should we make plots for those too?

**Comment by eurunuela** [comment]
I'm in favor of showing the data. In the end, it's just another button for us and some users may find it useful.

**Comment by jbteves** [comment]
I'm also in favor of showing them.

**Comment by tsalo** [comment]
Okay, sounds good. I'll come up with a list of other outputs we'll want carpet plots for, depending on the workflow parameters, and then add them soon.

**Comment by jbteves** [comment]
Let's stick with just these for now. Once this is merged and we see how users like it, I suspect we'll get some more feedback about what would be useful to include.

**Comment by tsalo** [comment]
Adding the new figures is pretty easy, but I'm not sure how to add the appropriate if/else statements to the HTML template. If I push the changes to add the figures, can someone else take over the HTML if statements?

**Comment by eurunuela** [comment]
> Adding the new figures is pretty easy, but I'm not sure how to add the appropriate if/else statements to the HTML template. If I push the changes to add the figures, can someone else take over the HTML if statements?

I'll take over the HTML. Just let me know when you're done ;)

File: tedana/reporting/data/html/report_body_template.html [hunk 1]
```diff
@@ -1,25 +1,173 @@
...
+  <h1>Carpet plots</h1>
+  <div class="carpet-buttons">
+    <button class="blue" type="button" onClick="updateCarpetPlot('optcom_nogsr')">
+      Optimally combined with global signal
+    </button>
+    <button class="blue" type="button" onClick="updateCarpetPlot('optcom')">
+      Optimally combined without global signal
+    </button>
+    <button class="yellow" type="button" onClick="updateCarpetPlot('denoised')">
+      Denoised before MIR
```
Comment (tsalo) [inline tedana/reporting/data/html/report_body_template.html:149]: I'd love it if this only said "before MIR" when MIR is run.

File: tedana/reporting/data/html/report_body_template.html [hunk 2]
```diff
@@ -1,25 +1,173 @@
...
+    </button>
+    <button class="yellow" type="button" onClick="updateCarpetPlot('denoised_mir')">
+      Denoised after MIR
+    </button>
+    <button class="green" type="button" onClick="updateCarpetPlot('accepted')">
+      Accepted before MIR
```
Comment (tsalo) [inline tedana/reporting/data/html/report_body_template.html:155]: I'd love it if this only said "before MIR" when MIR is run.

File: tedana/reporting/data/html/report_body_template.html [hunk 3]
```diff
@@ -1,25 +1,173 @@
...
+  $content
+</div>
+<div class="carpet-plots">
+  <h1>Carpet plots</h1>
+  <div class="carpet-buttons">
+    <button class="blue" type="button" onClick="updateCarpetPlot('optcom_nogsr')">
+      Optimally combined with global signal
+    </button>
+    <button class="blue" type="button" onClick="updateCarpetPlot('optcom')">
+      Optimally combined without global signal
```
Comment (tsalo) [inline tedana/reporting/data/html/report_body_template.html:146]: I'd love it if this only said "without global signal" when GSR is run.

**Comment by tsalo** [comment]
Thanks @eurunuela! It should be ready for you now. I also added comments to a few spots where it would be great to adjust the figure names based on whether GSR/MIR was applied.

**Comment by eurunuela** [comment]
So, what do you think @tsalo ? Is there anything else of the HTML you'd like changed?

I'm pretty happy with it. If anything, I'd play around until we got the right spacing between sections and stuff, but that's just me and my never-ending search for the perfect design.

**Comment by tsalo** [comment]
It's looking great, but would it be possible to have the button text vary based on whether GSR and/or MIR are run as well? Right now it says "Optimally combined without global signal", "Denoised before MIR", and "Accepted before MIR" even when GSR and MIR aren't run.

**Comment by eurunuela** [comment]
> It's looking great, but would it be possible to have the button text vary based on whether GSR and/or MIR are run as well? Right now it says "Optimally combined without global signal", "Denoised before MIR", and "Accepted before MIR" even when GSR and MIR aren't run.

I added a check so that MIR and GSR are only shown if the files are there. Do you have a better approach to achieve this?

**Comment by tsalo** [comment]
The figures themselves look great, including which figures actually show up, but the names in the buttons are the same regardless of which figures are included. Here are two screenshots to clarify:

## Four-echo test: GSR and MIR are both run, so there are buttons for the associated figures

image

## Five-echo test: No GSR or MIR, but the buttons mention both

image

If it's not straightforward to change the button titles based on which figures are available, then I'm fine switching back to using the following titles:

- `Optimally combined with global signal` (if GSR)
- `Optimally combined`
- `Denoised`
- `Denoised after MIR` (if MIR)
- `Accepted`
- `Accepted after MIR` (if MIR)
- `Rejected`

**Comment by eurunuela** [comment]
I understand now. I think that's doable. Let me give it a try.

**Comment by eurunuela** [comment]
There you go @tsalo 😉

**Comment by tsalo** [comment]
Thank you @eurunuela! It looks perfect.

File: tedana/io.py [hunk 4]
```diff
@@ -328,6 +328,51 @@ def add_decomp_prefix(comp_num, prefix, max_value):
...
+    mdata = data[mask]
+    dmdata = mdata.T - mdata.T.mean(axis=0)
+
+    # get variance explained by retained components
+    betas = get_coeffs(dmdata.T, mmix, mask=None)
+    varexpl = (1 - ((dmdata.T - betas.dot(mmix.T))**2.).sum() /
+               (dmdata**2.).sum()) * 100
+    LGR.info('Variance explained by decomposition: {:.02f}%'.format(varexpl))
+
+    # create component and de-noised time series and save to files
```
Comment (jbteves) [inline tedana/io.py:369]: Does this actually save to file here? I thought it just created the relevant time matrices.

Comment (tsalo) [reply tedana/io.py:369]: Good catch! It does not. I must have just copied the comment from the original version.

```suggestion
    # create component-based data
```

**Review by jbteves — APPROVED** [review]
LGTM, thanks @tsalo !

**Comment by tsalo** [comment]
@notZaki would you be willing to look at the HTML reports just to give your thoughts?

**Comment by notZaki** [comment]
That looks amazing!

I noticed that the plot tends to "bounce" whenever a button is pressed, and this can be distracting when switching between plots. This could be avoided by (1) making the buttons stay fixed when clicked by removing the `margin-top` style in:
```
button:active {
    ...
    margin-top: 3px;  /* Remove this line */
    ...
  }
```
or (2) add some padding below the buttons so that they don't shift the plot, by:
```
.carpet-buttons {
    height: 40px;
    padding-bottom: 5px;  /* Add this line */
  }
```

**Comment by eurunuela** [comment]
> That looks amazing!
> 
> 
> 
> I noticed that the plot tends to "bounce" whenever a button is pressed, and this can be distracting when switching between plots. This could be avoided by either (1) making the buttons stay fixed when clicked by removing the `margin-top` style in:
> 
> ```
> 
> button:active {
> 
>     ...
> 
>     margin-top: 3px;  /* Remove this line */
> 
>     ...
> 
>   }
> 
> ```
> 
> or (2) add some padding below the buttons so that they don't shift the plot, by:
> 
> ```
> 
> .carpet-buttons {
> 
>     height: 40px;
> 
>     padding-bottom: 5px;  /* Add this line */
> 
>   }
> 
> ```

Thank you @notZaki ! I'll all those!

**Comment by eurunuela** [comment]
I added @notZaki 's option 2. Thanks for the suggestion!

I will leave merging to you @tsalo .

**Review by eurunuela — APPROVED** [review]

**Pull request merged**